homeassistant/components/vicare/__init__.py

```diff
@@ -86,11 +94,32 @@ def setup(hass, config):
         return False
 
     hass.data[DOMAIN] = {}
+    hass.data[DOMAIN]["entities"] = {}
+    hass.data[DOMAIN]["entities"]["climate"] = []
     hass.data[DOMAIN][VICARE_API] = vicare_api
     hass.data[DOMAIN][VICARE_NAME] = conf[CONF_NAME]
     hass.data[DOMAIN][VICARE_HEATING_TYPE] = heating_type
 
     for platform in VICARE_PLATFORMS:
         discovery.load_platform(hass, platform, DOMAIN, {}, config)
 
+    def service_set_vicare_mode(service):
+        """Dispatch service calls to target entities."""
+        cmd = service.data[ATTR_COMMAND]
+        entity_id = service.data[ATTR_ENTITY_ID]
+        target_devices = [
+            dev
+            for dev in hass.data[DOMAIN]["entities"]["climate"]
+            if dev.entity_id in entity_id
+        ]
+
+        for target_device in target_devices:
+            target_device.set_vicare_mode(cmd)
+
+    hass.services.register(
```

Comment:
Use our entity service helper instead.

https://developers.home-assistant.io/docs/dev_101_services#entity-services

Reply:
seems like this is an async-only function.
I was trying to migrate the whole integration to async but the PyVicare library that I'm using not async-enabled.
Are there any good examples on how to deal with these kinds of libraries in an async integration?

Reply:
Use `hass.async_add_executor_job` to schedule sync tasks on the thread pool.

Reply:
not sure if I understood it correctly. Please have a look if I'm using it in the intended way. Thanks!